Extensions/CompositeMetadata.md

```diff
@@ -0,0 +1,43 @@
+# Composite Metadata Extension
+
+_This extension specification is currently incubating.  While incubating the version is 0._
+
+## Introduction
+There are a number of situations where an arbitrary collection of discrete metadata types should be attached to frame.  For example, a request frame may want to include both routing metadata as well as tracing metadata.  This extension specification provides an interoperable structure for metadadata payloads to contain multiple discrete metadata types.  It is designed such that if a consumer of the metadata is unaware of a particular type, it can be safely skipped and the next one read.
+
+## Metadata Payload
+This metadata type **MUST** only be used in frame types used to initiate interactions.  This includes [`REQUEST_FNF`][rf], [`REQUEST_RESPONSE`][rr], [`REQUEST_STREAM`][rs], and [`REQUEST_CHANNEL`][rc].  Multiple metadata payloads with the same MIME type are allowed.  The order of metadata payloads MUST be preserved when presented to responders.  The [`SETUP` Frame][s] Metadata MIME Type is `message/x.rsocket.composite-metadata.v0` and this metadata type has a well known id of `1`.
+
+[rc]: ../Protocol.md#frame-request-channel
+[rf]: ../Protocol.md#frame-fnf
+[rr]: ../Protocol.md#frame-request-response
+[rs]: ../Protocol.md#frame-request-stream
+[s]:  ../Protocol.md#frame-setup
+
+### Metadata Contents
+```
+     0                   1                   2                   3
+     0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    |M| MIME Length |   Metadata Encoding MIME Type                ...
+    +---------------+-------------------------------+---------------+
+    |              Metadata Length                  |
+    +-----------------------------------------------+---------------+
+    |                     Metadata Payload                         ...
+    +---------------+-----------------------------------------------+
+    |M| MIME Length |   Metadata Encoding MIME Type                ...
+    +---------------+-------------------------------+---------------+
+    |              Metadata Length                  |
+    +-----------------------------------------------+---------------+
+    |                     Metadata Payload                         ...
+    +---------------------------------------------------------------+
+                                   ...
+```
+
+* **Metadata Payload**: Any number of complete metadata payloads.
+  * (**M**)etadata Type: Metadata type is a well known value represented by a unique integer.
+  * **MIME Length**: (7 bits = max value 2^7 = 128) Unsigned 7-bit integer of the encoding MIME Type Length in bytes.  (Assumed to be 1 if M flag is set)
```

Comment:
If M is set, would the "Metadata Encoding MIME Type" be of length 0? and the rest of "MIME Length" is the value?  The "Assumed to be 1 if M flag is set" confuses me.

Reply:
@yschimke I would expect the MIME type to implicitly be an `byte`, and therefore the length is 1 byte.  From there the MIME type itself is just a unsigned byte value instead of a string.

Reply:
OK, I had assumed this was all going to be packed in a single byte with 127 magic values, instead of two bytes (-1 flag + 1 byte length, then the 1 byte with 255 values)

Reply:
Well, given @robertroeser's desire to support all of the well known MIME types, 127 isn't going to be enough options.

Reply:
I don't consider numbering all well known mime types to be viable given its a mutable set.  I hope you decide to just list ones that are widely applicable.

Reply:
@yschimke Another rev submitted.  Length is not included if M flag is set.